### PR TITLE
solvePnPRansac now accepts a seed and is deterministic

### PR DESCRIPTION
**Pull request by vpradeep07**
Currently, `solvePnPRansac` produces a different result each time it is called with the exact same input.  This non-determinism is problematic when integrating `solvePnPRansac` into systems that would otherwise have been deterministic.  This PR modifies `solvePnPRansac` to now produce deterministic results.

The changes include:
- Passing in a seed for the `RNG`
- Breaking ties in the number of inliers by picking the iteration with a later index
- Rerunning the final solvePnP (containing all the inliers) without an extrinsics guess
- Fixing a race condition, so that the inliers count is not modified while inliers are being modified (also seen in PR #3587)
- Ensuring that the 4 initial points chosen for a given iteration are not affected by the previous iteration

Since this PR modifies the function signature of `solvePnPRansac`, it results in an ABI change.  If this is problematic, we could provide an overloaded version of `solvePnPRansac` that takes the seed.

### CODE REVIEW

**Comment by vpradeep07** [comment]
The build error above seems to be due to spurious error on the build server.  How can I trigger a rebuild?

Will the following trigger a rebuild?

test this please

**Comment by ilya-lavrenov** [comment]
> Will the following trigger a rebuild?

Changing of last commit hash.

File: modules/calib3d/include/opencv2/calib3d/calib3d.hpp [hunk 1]
```diff
@@ -490,6 +490,21 @@ CV_EXPORTS_W void solvePnPRansac( InputArray objectPoints,
                                   OutputArray inliers = noArray(),
                                   int flags = ITERATIVE);
 
+//! computes the camera pose from a few 3D points and the corresponding projections. The outliers are possible.
+CV_EXPORTS_W void solvePnPRansac( InputArray objectPoints,
```
Comment (ilya-lavrenov) [inline modules/calib3d/include/opencv2/calib3d/calib3d.hpp]: don't introduce another function, just add an optional parameter to the original one.

Comment (vpradeep07) [inline modules/calib3d/include/opencv2/calib3d/calib3d.hpp]: Thanks for the comment.  It is now addressed

**Comment by vpisarev** [comment]
@alekcac, this is another patch in SolvePnPRansac; could you please review it too?

**Comment by alekcac** [comment]
Ok, I will review it at the weekend.

**Comment by alekcac** [comment]
Sorry for making you wait, I was ill some time. In common I agree with these changes. But I didn't understand one moment. Could you please comment it before merge.

File: modules/calib3d/src/solvepnp.cpp [hunk 2]
```diff
@@ -357,7 +359,7 @@ void cv::solvePnPRansac(InputArray _opoints, InputArray _ipoints,
                 Mat colInlierObjectPoints = inlierObjectPoints(Rect(i, 0, 1, 1));
                 objectPoints.col(index).copyTo(colInlierObjectPoints);
             }
-            solvePnP(inlierObjectPoints, inlierImagePoints, params.camera.intrinsics, params.camera.distortion, localRvec, localTvec, true, flags);
+            solvePnP(inlierObjectPoints, inlierImagePoints, params.camera.intrinsics, params.camera.distortion, localRvec, localTvec, false, flags);
```
Comment (alekcac) [inline modules/calib3d/src/solvepnp.cpp]: What is the reason of this change? In my opinion we should use true option to use previous hypothesis about pose.

Comment (vpradeep07) [inline modules/calib3d/src/solvepnp.cpp]: In `PnPSolver::operator()`, if a sample/iteration has more inliers than the minimum inlier count (`if ((int)inliers.size() >= parameters.minInliersCount)`), then we terminate early.  Thus, if there are two samples that potentially have more inliers than `minInliersCount`, there is a race condition as to which sample will be chosen, yielding different results.

One might guess that this issue is resolved by setting minInliersCount to -1, which requires all points to be inliers in order to exit early.  In this situation, the `localInliers` vector will be the same for both samples, but the `localRvec` and `localTvec` variables will differ.  Thus, if `localRVec` and `localTvec` are then used to compute the final `solvePnP` solution, we will have a non-determinism.  Setting the `useExtrinsicGuess` option to false avoids the use of `localRvec` and `localTvec` in the final `solvePnP` solve, and thus keeps the algorithm deterministic when `minInliersCount` is `-1`.

**Comment by vpisarev** [comment]
@vpradeep07, thanks for the patch! In 2.4 we can not modify API at all, e.g. add parameters even if the parameters are optional. Please, remove rng_seed from parameters, make it internal constant in the function. Otherwise we can not accept the patch, sorry.

File: modules/calib3d/src/solvepnp.cpp (hunk 2 above)
Comment (alekcac) [inline modules/calib3d/src/solvepnp.cpp]: Ok, but as I understand, it doesn't solve the problem with random results for minInliersCount >= 1
Is it so?

**Comment by alekcac** [comment]
Ok, I agree with these changes. We can merge it as soon as remove issue with ABI. Vadim, could you please comment previous message about ABI compatibility?

**Comment by vpisarev** [comment]
@vpradeep07, you are right, different OpenCV 2.4.x must be binary compatible, otherwise OpenCV Manager on Android will crash; and there may be some other bad consequences that we are not aware of. I would like to avoid adding overloaded signature, I guess, exposing rng_seed is desirable, but not must-have feature. So, yes, I kindly ask you to just hide it inside. In the master branch (to become OpenCV 3.0) we can add this optional rng_seed easily.

**Comment by alekcac** [comment]
@vpradeep07 Any updates?

**Comment by vpradeep07** [comment]
@alekcac - Sorry for the delay here.  I made the appropriate changes and the build passed, so we should be good to merge.

**Comment by vpisarev** [comment]
@alekcac, do you approve the PR?

**Comment by alekcac** [comment]
:+1: @vpradeep07 Thanks a lot for the request!
